Error running self-contained Linux EXE: "This executable is not bound to a managed DLL to execute. The binding value is: 'c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2'"

I created a new console application in VS 2017 using .NET Core 2.1. Published it as a self-contained linux-x64 app without any changes and tried to run it on an Ubuntu 18.04 VM. Got this error:

```
 ./LinuxTest 
This executable is not bound to a managed DLL to execute. The binding value is: 'c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2'
A fatal error was encountered. This executable was not bound to load a managed DLL.
```

c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2 is apparently the SHA256 hash of "foobar", so whatever the real problem is, this is a very weird error message, too!

LinuxTest.csproj:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>

</Project>
```

Program.cs:
```
using System;

namespace LinuxTest
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }
}
```

Resulting self-contained binary: [LinuxTest.zip](https://github.com/dotnet/sdk/files/2900841/LinuxTest.zip)

If I install dotnet-runtime-2.1 on the Linux VM and run `dotnet LinuxTest.dll` it works.

OK, so the repro is a little more complicated! I actually have 2 Linux VM on my Windows host: a Linux Mint 19 one with the .NET Core SDK installed and an Ubuntu 18.04 one, without .NET Core installed. The Windows host also has .NET Core SDK installed. (dotnet --info for both attached.)

I initially tried to publish on the Mint VM, got an error, re-ran the same command and it appeared to work:

```
evgeny@emmint:/media/sf_work/Play/LinuxTest$ dotnet publish -r linux-x64 -o publish-linux-from-mint
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for /media/sf_work/Play/LinuxTest/LinuxTest.csproj...
  Generating MSBuild file /media/sf_work/Play/LinuxTest/obj/LinuxTest.csproj.nuget.g.props.
  Generating MSBuild file /media/sf_work/Play/LinuxTest/obj/LinuxTest.csproj.nuget.g.targets.
  Restore completed in 476.51 ms for /media/sf_work/Play/LinuxTest/LinuxTest.csproj.
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018: The "EmbedAppNameInHost" task failed unexpectedly. [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018: System.IO.IOException: Invalid argument [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at System.IO.MemoryMappedFiles.MemoryMappedView.CreateView(SafeMemoryMappedFileHandle memMappedFileHandle, MemoryMappedFileAccess access, Int64 requestedOffset, Int64 requestedSize) [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at System.IO.MemoryMappedFiles.MemoryMappedFile.CreateViewAccessor(Int64 offset, Int64 size, MemoryMappedFileAccess access) [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at System.IO.MemoryMappedFiles.MemoryMappedFile.CreateViewAccessor() [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at Microsoft.NET.Build.Tasks.AppHost.Create(String appHostSourceFilePath, String appHostDestinationFilePath, String appBinaryFilePath, Boolean overwriteExisting) [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at Microsoft.NET.Build.Tasks.EmbedAppNameInHost.ExecuteCore() [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]
/usr/share/dotnet/sdk/2.1.503/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(308,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/media/sf_work/Play/LinuxTest/LinuxTest.csproj]

evgeny@emmint:/media/sf_work/Play/LinuxTest$ dotnet publish -r linux-x64 -o publish-linux-from-mint
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 64.45 ms for /media/sf_work/Play/LinuxTest/LinuxTest.csproj.
  LinuxTest -> /media/sf_work/Play/LinuxTest/bin/Debug/netcoreapp2.1/linux-x64/LinuxTest.dll
  LinuxTest -> /media/sf_work/Play/LinuxTest/publish-linux-from-mint/
```

(`/media/sf_work` here is the host NTFS directory `C:\W ork` mounted inside the VM)

However, when I run the output I get the error int he original issue.

I then tried to publish on the Windows host:

```
C:\W ork\Play\LinuxTest>dotnet publish -r linux-x64 -o publish-linux-from-windows-cmd
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for C:\W ork\Play\LinuxTest\LinuxTest.csproj...
  Generating MSBuild file C:\W ork\Play\LinuxTest\obj\LinuxTest.csproj.nuget.g.props.
  Generating MSBuild file C:\W ork\Play\LinuxTest\obj\LinuxTest.csproj.nuget.g.targets.
  Restore completed in 233.6 ms for C:\W ork\Play\LinuxTest\LinuxTest.csproj.
  LinuxTest -> C:\W ork\Play\LinuxTest\bin\Debug\netcoreapp2.1\linux-x64\LinuxTest.dll
  LinuxTest -> C:\W ork\Play\LinuxTest\publish-linux-from-windows-cmd\
```

This appears to succeed, but the resulting executable gives the same error (even though I'm publishing to a different directory!) Even if I run `dotnet clean` on the Windows host and try to publish again it keeps happening. Only if I delete the `obj` directory and publish again does it work!

So it seems like there is a number of bugs here:

1) The original `The "EmbedAppNameInHost" task failed unexpectedly` error
2) The fact that running publish a second time in the Mint VM gives no error, but produces a non-working EXE.
3) That this affects future attempts to publish, even to a different directory.
4) That even `dotnet clean` does not actually cause `dotnet publish` to start from a clean state, as the user would expect.

[dotnet-info-Linux-Mint.txt](https://github.com/dotnet/sdk/files/2905333/dotnet-info-Linux-Mint.txt)
[dotnet-info-Windows.txt](https://github.com/dotnet/sdk/files/2905334/dotnet-info-Windows.txt)

--- a/src/Tasks/Common/Resources/Strings.resx
+++ b/src/Tasks/Common/Resources/Strings.resx
@@ -540,12 +540,16 @@ The following are names of parameters or literal values and should not be transl
     <value>NETSDK1108: ReadyToRun compilation will be skipped because it is only supported for executable projects.</value>
     <comment>{StrBegin="NETSDK1108: "}</comment>
   </data>
-    <data name="RuntimeListNotFound" xml:space="preserve">
+  <data name="RuntimeListNotFound" xml:space="preserve">
     <value>NETSDK1109: Runtime list file '{0}' was not found. Report this error to the .NET Core team here: https://aka.ms/dotnet-sdk-issue.</value>
     <comment>{StrBegin="NETSDK1109: "}</comment>
   </data>
   <data name="DuplicateRuntimePackAsset" xml:space="preserve">
     <value>NETSDK1110: More than one asset in the runtime pack has the same destination sub-path of '{0}'. Report this error to the .NET Core team here: https://aka.ms/dotnet-sdk-issue.</value>
     <comment>{StrBegin="NETSDK1110: "}</comment>
+  </data>
+  <data name="FailedToDeleteApphost" xml:space="preserve">
+    <value>NETSDK1111: Failed to delete output apphost: {0}</value>
+    <comment>{StrBegin="NETSDK1111: "}</comment>
   </data>
 </root>

--- a/src/Tasks/Common/Resources/xlf/Strings.cs.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.cs.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: Chyba při načítání souboru prostředků: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: Nepovedlo se uzamknout prostředek.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.de.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.de.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: Fehler beim Lesen der Ressourcendatei: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: Fehler beim Sperren der Ressource.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.es.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.es.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: Error al leer el archivo de activos: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: No se pudo bloquear el recurso.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.fr.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.fr.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: Erreur durant la lecture du fichier de composants : {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: échec du verrouillage de la ressource.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.it.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.it.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: errore durante la lettura del file di asset: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: non è stato possibile bloccare la risorsa.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.ja.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.ja.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: アセット ファイルの読み取りでエラーが発生しました: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: リソースをロックできませんでした。</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.ko.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.ko.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: 자산 파일 {0}을(를) 읽는 동안 오류가 발생했습니다.</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: 리소스를 잠그지 못했습니다.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.pl.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.pl.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: Błąd podczas odczytywania pliku zasobów: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: Nie można zablokować zasobu.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: Erro ao ler o arquivo de ativos: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: falha ao bloquear o recurso.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.ru.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.ru.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: ошибка при чтении файла ресурсов {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: не удалось заблокировать ресурс.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.tr.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.tr.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: Varlıklar dosyası okunurken hata: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: Kaynak kilitlenemedi.</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.zh-Hans.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.zh-Hans.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: 读取资产文件时出错: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: 无法锁定资源。</target>

--- a/src/Tasks/Common/Resources/xlf/Strings.zh-Hant.xlf
+++ b/src/Tasks/Common/Resources/xlf/Strings.zh-Hant.xlf
@@ -267,6 +267,11 @@
         <target state="translated">NETSDK1060: 讀取資產檔案時發生錯誤: {0}</target>
         <note>{StrBegin="NETSDK1060: "}</note>
       </trans-unit>
+      <trans-unit id="FailedToDeleteApphost">
+        <source>NETSDK1111: Failed to delete output apphost: {0}</source>
+        <target state="new">NETSDK1111: Failed to delete output apphost: {0}</target>
+        <note>{StrBegin="NETSDK1111: "}</note>
+      </trans-unit>
       <trans-unit id="FailedToLockResource">
         <source>NETSDK1077: Failed to lock resource.</source>
         <target state="translated">NETSDK1077: 無法鎖定資源。</target>

--- a/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAnAppHost.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAnAppHost.cs
@@ -68,6 +68,8 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
                     .Contain(sourceAppHostMock)
                     .And
                     .Contain(AppBinaryPathPlaceholder);
+
+                File.Exists(destinationFilePath).Should().BeFalse();
             }
         }
 
@@ -88,6 +90,8 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
                     .Message
                     .Should()
                     .Contain(appBinaryFilePath);
+
+                File.Exists(destinationFilePath).Should().BeFalse();
             }
         }
 
@@ -161,6 +165,8 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
                     .Message
                     .Should()
                     .Contain(sourceAppHostMock);
+
+                File.Exists(destinationFilePath).Should().BeFalse();
             }
         }
 

--- a/src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs
@@ -52,45 +52,54 @@ namespace Microsoft.NET.Build.Tasks
             // Copy apphost to destination path so it inherits the same attributes/permissions.
             File.Copy(appHostSourceFilePath, appHostDestinationFilePath, overwrite: true);
 
-            // Re-write the destination apphost with the proper contents.
-            bool appHostIsPEImage = false;
-            using (var memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostDestinationFilePath))
+            try
             {
-                using (MemoryMappedViewAccessor accessor = memoryMappedFile.CreateViewAccessor())
+                // Re-write the destination apphost with the proper contents.
+                bool appHostIsPEImage = false;
+                using (var memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostDestinationFilePath))
                 {
-                    SearchAndReplace(accessor, AppBinaryPathPlaceholderSearchValue, bytesToWrite, appHostSourceFilePath);
-
-                    appHostIsPEImage = IsPEImage(accessor);
-
-                    if (windowsGraphicalUserInterface)
+                    using (MemoryMappedViewAccessor accessor = memoryMappedFile.CreateViewAccessor())
                     {
-                        if (!appHostIsPEImage)
-                        {
-                            throw new BuildErrorException(Strings.AppHostNotWindows, appHostSourceFilePath);
-                        }
+                        SearchAndReplace(accessor, AppBinaryPathPlaceholderSearchValue, bytesToWrite, appHostSourceFilePath);
 
-                        SetWindowsGraphicalUserInterfaceBit(accessor, appHostSourceFilePath);
+                        appHostIsPEImage = IsPEImage(accessor);
+
+                        if (windowsGraphicalUserInterface)
+                        {
+                            if (!appHostIsPEImage)
+                            {
+                                throw new BuildErrorException(Strings.AppHostNotWindows, appHostSourceFilePath);
+                            }
+
+                            SetWindowsGraphicalUserInterfaceBit(accessor, appHostSourceFilePath);
+                        }
                     }
                 }
-            }
 
-            if (intermediateAssembly != null && appHostIsPEImage)
+                if (intermediateAssembly != null && appHostIsPEImage)
+                {
+                    if (ResourceUpdater.IsSupportedOS())
+                    {
+                        // Copy resources from managed dll to the apphost
+                        new ResourceUpdater(appHostDestinationFilePath)
+                            .AddResourcesFromPEImage(intermediateAssembly)
+                            .Update();
+                    }
+                    else if (log != null)
+                    {
+                        log.LogWarning(Strings.AppHostCustomizationRequiresWindowsHostWarning);
+                    }
+                }
+
+                // Memory-mapped write does not updating last write time
+                File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+            }
+            catch (Exception)
             {
-                if (ResourceUpdater.IsSupportedOS())
-                {
-                    // Copy resources from managed dll to the apphost
-                    new ResourceUpdater(appHostDestinationFilePath)
-                        .AddResourcesFromPEImage(intermediateAssembly)
-                        .Update();
-                }
-                else if (log != null)
-                {
-                    log.LogWarning(Strings.AppHostCustomizationRequiresWindowsHostWarning);
-                }
+                // Delete the destination file so we don't leave an unmodified apphost
+                File.Delete(appHostDestinationFilePath);
+                throw;
             }
-
-            // Memory-mapped write does not updating last write time
-            File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
         }
 
         // See: https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm

--- a/src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs
@@ -97,7 +97,14 @@ namespace Microsoft.NET.Build.Tasks
             catch (Exception)
             {
                 // Delete the destination file so we don't leave an unmodified apphost
-                File.Delete(appHostDestinationFilePath);
+                try
+                {
+                    File.Delete(appHostDestinationFilePath);
+                }
+                catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException)
+                {
+                    log?.LogError(Strings.FailedToDeleteApphost, ex.Message);
+                }
                 throw;
             }
         }

